cdap-api/src/main/java/co/cask/cdap/api/app/Application.java

```diff
@@ -16,16 +16,20 @@
 
 package co.cask.cdap.api.app;
 
+import co.cask.cdap.api.templates.plugins.PluginConfig;
+
 /**
  * Defines a CDAP Application.
  *
+ * @param <T> {@link PluginConfig} class that represents the configuration of the Application.
+ *
  */
-public interface Application {
+public interface Application<T extends PluginConfig> {
```

Comment:
I think it's better to have a better naming for `PluginConfig` first (you can have an new interface and have `PluginConfig` to extend from it with empty body for backward compatibility).